sstring: fix format and optimize the performance of sstring::find().

balusch opened the pull request:
This patch optimize sstring::find() in two ways:

- sstring::find used to compare character one by one. The patch makes
sstring::find to get converted to calls to memchr and memcmp, which is
much more efficient.
- This patch will stop seaching earlier when source string is shorter
than pattern string to avoid unnecessary comparison.

This patch is ported from the implementation of std::string::find in LLVM:
https://github.com/llvm/llvm-project/blob/main/libcxx/include/__string#L939

psarna commented:
Did you run any benchmarks comparing the new code against the new one? Looks very similar to https://github.com/llvm/llvm-project/blob/main/libcxx/include/__string#L939 indeed, so I'm fairly confident about the correctness, but I'd appreciate a second opinion anyway, since this code is quite essential. @nyh maybe?

nyh commented on include/seastar/core/sstring.hh:
```diff
@@ -243,28 +243,40 @@ public:
+        }
+
+        char_type f2 = *c_str;
+        do {
```
nitpick: do/while is an esoteric replacement for the normal while() that is useful in some cases (when the condition doesn't need to be evaluated on the first run) but not in this code, where a normal while() would have sufficed.

balusch replied:
Yes, the LLVM's version use the normal `while()` loop, but since we have already calculated `len1` and ensured that `len1 < len2` before the loop, we can just skip the length calculation and the comparison in the first loop cycle, so I replace it with `do-while` loop.

nyh reviewed (commented):
The patch looks correct to me, but it's a risky patch (changing something so low-level) so let's at least document in the commit message the motivation of this patch. What kind of performance improvement did you measure here?

balusch commented:
> Did you run any benchmarks comparing the new code against the new one? 

@psarna just copied the benchmark code written in reviews.llvm.org/D27068 and modified it for `seastar::sstring`(I should have ran it before the PR 😅). Since I haven't seen any benchmark framework used in Seastar, just some unit tests and perf code so I paste the code here in case that the code won't compile(Or maybe I can add [google/benchmark](https://github.com/google/benchmark) support for Seastar?)

```c++
#include <seastar/core/sstring.hh>
#include <benchmark/benchmark.h>

constexpr std::size_t MAX_STRING_LEN = 8 << 14;

// Benchmark when there is no match.
static void BM_StringFindNoMatch(benchmark::State &state) {
  seastar::sstring s1(state.range(0), '-');
  seastar::sstring s2(8, '*');
  while (state.KeepRunning())
    benchmark::DoNotOptimize(s1.find(s2));
}
BENCHMARK(BM_StringFindNoMatch)->Range(10, MAX_STRING_LEN);

// Benchmark when the string matches first time.
static void BM_StringFindAllMatch(benchmark::State &state) {
  seastar::sstring s1(MAX_STRING_LEN, '-');
  seastar::sstring s2(state.range(0), '-');
  while (state.KeepRunning())
    benchmark::DoNotOptimize(s1.find(s2));
}
BENCHMARK(BM_StringFindAllMatch)->Range(1, MAX_STRING_LEN);

// Benchmark when the string matches somewhere in the end.
static void BM_StringFindMatch1(benchmark::State &state) {
  seastar::sstring s1(MAX_STRING_LEN / 2, '*');
  s1 += seastar::sstring(state.range(0), '-');
  seastar::sstring s2(state.range(0), '-');
  while (state.KeepRunning())
    benchmark::DoNotOptimize(s1.find(s2));
}
BENCHMARK(BM_StringFindMatch1)->Range(1, MAX_STRING_LEN / 4);

// Benchmark when the string matches somewhere from middle to the end.
static void BM_StringFindMatch2(benchmark::State &state) {
  seastar::sstring s1(MAX_STRING_LEN / 2, '*');
  s1 += seastar::sstring(state.range(0), '-');
  s1 += seastar::sstring(state.range(0), '*');
  seastar::sstring s2(state.range(0), '-');
  while (state.KeepRunning())
    benchmark::DoNotOptimize(s1.find(s2));
}
BENCHMARK(BM_StringFindMatch2)->Range(1, MAX_STRING_LEN / 4);

BENCHMARK_MAIN();
```

And here is the result:

master:
```text
Run on (8 X 3792 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 256 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 2.94, 6.31, 4.83
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
BM_StringFindNoMatch/10            7.43 ns         7.40 ns     94319503
BM_StringFindNoMatch/64            46.2 ns         46.0 ns     15206199
BM_StringFindNoMatch/512            338 ns          338 ns      2056814
BM_StringFindNoMatch/4096          2754 ns         2738 ns       245829
BM_StringFindNoMatch/32768        21305 ns        21238 ns        31901
BM_StringFindNoMatch/131072       87668 ns        87338 ns         8171
BM_StringFindAllMatch/1            2.66 ns         2.65 ns    262279932
BM_StringFindAllMatch/8            5.35 ns         5.34 ns    132248558
BM_StringFindAllMatch/64           29.3 ns         29.2 ns     23666330
BM_StringFindAllMatch/512           227 ns          227 ns      3090288
BM_StringFindAllMatch/4096         1768 ns         1764 ns       395518
BM_StringFindAllMatch/32768       14148 ns        14111 ns        49597
BM_StringFindAllMatch/131072      56426 ns        56283 ns        12045
BM_StringFindMatch1/1             42379 ns        42273 ns        16671
BM_StringFindMatch1/8             42471 ns        42371 ns        16574
BM_StringFindMatch1/64            42453 ns        42336 ns        16531
BM_StringFindMatch1/512           42372 ns        42273 ns        16515
BM_StringFindMatch1/4096          43658 ns        43552 ns        16059
BM_StringFindMatch1/32768         52923 ns        52792 ns        13100
BM_StringFindMatch2/1             42158 ns        42052 ns        16578
BM_StringFindMatch2/8             42347 ns        42244 ns        16609
BM_StringFindMatch2/64            42330 ns        42228 ns        16502
BM_StringFindMatch2/512           42897 ns        42793 ns        16410
BM_StringFindMatch2/4096          44453 ns        44332 ns        15785
BM_StringFindMatch2/32768         57906 ns        57749 ns        11947
```

master + patch:
```text
Run on (8 X 3792 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 256 KiB (x8)
  L3 Unified 16384 KiB (x1)
Load Average: 5.30, 7.18, 5.03
-----------------------------------------------------------------------
Benchmark                             Time             CPU   Iterations
-----------------------------------------------------------------------
BM_StringFindNoMatch/10            2.84 ns         2.83 ns    223265993
BM_StringFindNoMatch/64            3.43 ns         3.42 ns    209920608
BM_StringFindNoMatch/512           6.60 ns         6.58 ns    107157634
BM_StringFindNoMatch/4096          24.8 ns         24.7 ns     28192009
BM_StringFindNoMatch/32768          187 ns          186 ns      3749372
BM_StringFindNoMatch/131072        1192 ns         1188 ns       590624
BM_StringFindAllMatch/1            5.11 ns         5.10 ns    135206314
BM_StringFindAllMatch/8            5.26 ns         5.24 ns    130680897
BM_StringFindAllMatch/64           5.00 ns         4.98 ns    141131717
BM_StringFindAllMatch/512          11.0 ns         11.0 ns     64716153
BM_StringFindAllMatch/4096         47.0 ns         46.9 ns     14861469
BM_StringFindAllMatch/32768         768 ns          766 ns       889502
BM_StringFindAllMatch/131072       3393 ns         3384 ns       207499
BM_StringFindMatch1/1               532 ns          531 ns      1254398
BM_StringFindMatch1/8               534 ns          533 ns      1298437
BM_StringFindMatch1/64              541 ns          540 ns      1261582
BM_StringFindMatch1/512             543 ns          541 ns      1279840
BM_StringFindMatch1/4096            603 ns          602 ns      1144986
BM_StringFindMatch1/32768          1244 ns         1241 ns       561752
BM_StringFindMatch2/1               652 ns          650 ns      1077974
BM_StringFindMatch2/8               536 ns          535 ns      1295882
BM_StringFindMatch2/64              537 ns          536 ns      1264378
BM_StringFindMatch2/512             545 ns          543 ns      1221322
BM_StringFindMatch2/4096            629 ns          627 ns      1086925
BM_StringFindMatch2/32768          1295 ns         1289 ns       523033
```

nyh commented:
Very nice benchmark results! A really impressive speedup for long strings (it seems this benchmark only uses long ~32 KB strings as the string to search in, right?)

balusch commented:
> it seems this benchmark only uses long ~32 KB strings as the string to search in, right?

The source string can be as long as 128KB or as short as a few bytes.

balusch commented:
> so let's at least document in the commit message the motivation of this patch.

@nyh I added some comment in the commit message and replaced `do-while` with normal `while()` loop, please have a look again. Thanks!
